### PR TITLE
[CSL-177] Do not accept block which arrived after slot finish

### PR DESCRIPTION
**Pull request by neongreen**
I haven't run it yet because I don't know how to connect to testnet. (Will fix merge conflicts after I find out how to run it.)

I spent 90% of time trying to understand how recovery mode works, and 10% of time – writing code. Correspondingly, half of this PR consists of adding comments and renaming things. If you want to see changes to logic, read https://github.com/input-output-hk/cardano-sl/commit/d8f73015f2a02e7c5bf2595c50c49a40a892a92b.

### CODE REVIEW

File: src/Pos/Block/Network/Retrieval.hs [hunk 1]
```diff
@@ -57,68 +57,103 @@ retrievalWorker getPeers = worker outs (retrievalWorkerImpl getPeers)
...
 
+-- I really don't like join
+{-# ANN retrievalWorkerImpl ("HLint: ignore Use join" :: Text) #-}
+
+-- | Worker that queries blocks. It has two jobs:
+--
+-- * If there are headers in 'ncBlockRetrievalQueue', this worker retrieves
+--   blocks according to that queue.
+--
+-- * If recovery is in progress, this worker keeps recovery going.
```
Comment (volhovM) [inline src/Pos/Block/Network/Retrieval.hs]: ... by asking headers and backing up to (1 ~ block retrieval) on the next loop iteration. Please add.

File: src/Pos/Block/Network/Retrieval.hs [hunk 2]
```diff
@@ -57,68 +57,103 @@ retrievalWorker getPeers = worker outs (retrievalWorkerImpl getPeers)
...
-        whenJustM (mkHeadersRequest (Just $ headerHash rHeader)) $ \mghNext ->
-            handleAll (handleHeadersRecoveryE peerId) $
-            reportingFatal getPeers version $
-            reifyMsgLimit (Proxy @(MsgHeaders ssc)) $ \limPx ->
-            withConnectionTo sendActions peerId $ \_peerData ->
-                requestHeaders mghNext peerId (Just rHeader) limPx
+    mainLoopE e = do
+        logError $ sformat ("retrievalWorker: error caught "%shown) e
+        throw e
+    --
```
Comment (volhovM) [inline src/Pos/Block/Network/Retrieval.hs]: Zabyl.

File: src/Pos/Block/Network/Retrieval.hs [hunk 3]
```diff
@@ -57,68 +57,103 @@ retrievalWorker getPeers = worker outs (retrievalWorkerImpl getPeers)
...
+        handleAll (handleBlockRetrievalE chunk) $
+        reportingFatal getPeers version $
+        workerHandle getPeers sendActions chunk
     handleBlockRetrievalE (peerId, headers) e = do
         logWarning $ sformat
             ("Error handling peerId="%build%", headers="%listJson%": "%shown)
             peerId (fmap headerHash headers) e
         dropUpdateHeader
         dropRecoveryHeaderAndRepeat getPeers sendActions peerId
+    --
```
Comment (volhovM) [inline src/Pos/Block/Network/Retrieval.hs]: Mystery.

File: src/Pos/Block/Network/Retrieval.hs [hunk 4]
```diff
@@ -57,68 +57,103 @@ retrievalWorker getPeers = worker outs (retrievalWorkerImpl getPeers)
...
 dropUpdateHeader :: WorkMode ssc m => m ()
 dropUpdateHeader = do
     progressHeaderVar <- ncProgressHeader <$> getNodeContext
     void $ atomically $ tryTakeTMVar progressHeaderVar
 
-dropRecoveryHeader :: WorkMode ssc m => NodeId -> m Bool
+dropRecoveryHeader
+    :: WorkMode ssc m
+    => NodeId
+    -> m Bool           -- ^ Whether given peer was kicked and recovery was
```
Comment (volhovM) [inline src/Pos/Block/Network/Retrieval.hs]: Move this comment to function comment instead. I guess it's also worth mentioning the reason `peerId` should be passed. 
Consider scenario when you work with peer P and header H on recovery mode, it fails, at the same time other thread (new block arrived) replaces header to H2 and peer to P2, we drop communication with H2 and halt (while in fact we should ignore this exception and continue to work with P2).

File: src/Pos/Block/Network/Retrieval.hs [hunk 5]
```diff
@@ -186,11 +227,21 @@ workerHandle getPeers sendActions (peerId, headers) = do
     uselessFormat =
         "Chain of headers from " %shortHashF % " to " %shortHashF %
         " is useless for the following reason: " %stext
+    oldFormat =
+        "Chain of headers from " %shortHashF % " to " %shortHashF %
+        " is invalid for the following reason: the newest header is from"%
```
Comment (volhovM) [inline src/Pos/Block/Network/Retrieval.hs]: Besides I've asked to move this message into `CHsUseless`, i'll emphasize once more: it's not invalid from the design perspective. Invalid stays for "broken" so we can effectively ban neighbor if we got something malformed from her. Useless info is just skipped.

File: src/Pos/Block/Network/Retrieval.hs (hunk 3 above)
Comment (gromakovsky) [inline src/Pos/Block/Network/Retrieval.hs]: He told me it's intentional.

File: src/Pos/Block/Network/Retrieval.hs (hunk 5 above)
Comment (neongreen) [inline src/Pos/Block/Network/Retrieval.hs]: Okay, will fix.

File: src/Pos/Block/Network/Logic.hs [hunk 6]
```diff
@@ -289,7 +282,7 @@ requestHeaders mgh peerId origTip _ conv = do
 -- First case of 'handleBlockheaders'
 handleRequestedHeaders
     :: forall ssc m.
-       (WorkMode ssc m)
+       (SscWorkersClass ssc, WorkMode ssc m)
```
Comment (neongreen) [inline src/Pos/Block/Network/Logic.hs]: I think GHC asked for it. Not sure. I'll check and if it's not needed, I'll remove it.

File: src/Pos/Block/Logic.hs [hunk 7]
```diff
@@ -222,37 +252,62 @@ classifyNewHeader (Right header) = do
...
+    needRecovery_ <- needRecovery
+    mbCurrentSlot <- getCurrentSlot
+    let newestHeaderConvertedSlot =
+            case newestHeader ^. epochOrSlotG of
+                EpochOrSlot (Left e)  -> SlotId e 0
+                EpochOrSlot (Right s) -> s
     if | not headersValid ->
-             pure $ CHsInvalid "Header chain is invalid"
+             pure $ CHsInvalid
+             "Header chain is invalid"
```
Comment (neongreen) [inline src/Pos/Block/Logic.hs]: I don't remember, probably for consistency. Or maybe there was a long string somewhere and then I deleted it and forgot to change things back. I'll revert it.

File: src/Pos/Block/Logic.hs [hunk 8]
```diff
@@ -147,6 +149,34 @@ lcaWithMainChain headers =
...
+    => m Bool
+needRecovery =
+    getCurrentSlot >>= \case
+        Nothing   -> pure True
+        Just slot -> isTooOld slot
+  where
+    isTooOld slot = do
+        knownSlot <- getEpochOrSlot <$> DB.getTipBlockHeader @ssc
+        pure $
+            (knownSlot & from enum %~ (+ blkSecurityParam)) <
```
Comment (neongreen) [inline src/Pos/Block/Logic.hs]: Eh, `+~` is weird and hard to see.

Comment (int-index) [inline src/Pos/Block/Logic.hs]: I think `+~` is just fine.

Comment (neongreen) [inline src/Pos/Block/Logic.hs]: I will clarify: _for me_ it is weird and hard to see. I.e. unless I look closely it looks like a weird `--`.

If someone feels strongly about it, I'll change it.

Comment (neongreen) [inline src/Pos/Block/Logic.hs]: > Also are you 100% sure converting to/from enum won't break the getEpochOrSlot?

Well, there is some semi-complicated instance for `Enum EpochOrSlot` (I think @gromakovsky wrote it), but I guess it's actually inappropriate in this case:

```
> fromEnum (EpochOrSlot (Left 1))
21
> fromEnum (EpochOrSlot (Right (SlotId 1 0)))
22
```

The “genesis slot” doesn't take any time so it shouldn't be counted in `isTooOld`. I'll fix it.

File: src/Pos/Block/Logic.hs [hunk 9]
```diff
@@ -222,37 +252,62 @@ classifyNewHeader (Right header) = do
...
              "Didn't manage to find block corresponding to parent " <>
              "of oldest element in chain (should be one of checkpoints)"
        | newestHash == headerHash tip ->
-             pure $ CHsUseless "Newest hash is the same as our tip"
+             pure $ CHsUseless
+             "Newest hash is the same as our tip"
        | newestHeader ^. difficultyL <= tipHeader ^. difficultyL ->
-             pure $ CHsUseless "Newest hash difficulty is not greater than our tip's"
+             pure $ CHsUseless
+             "Newest hash difficulty is not greater than our tip's"
```
Comment (rockbmb) [inline src/Pos/Block/Logic.hs]: I guess here and in the new line 303 the string can be moved up.

**Pull request merged**